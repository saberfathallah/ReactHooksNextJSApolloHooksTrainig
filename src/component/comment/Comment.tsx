import React from "react";
import Avatar from "@material-ui/core/Avatar";
import ListItemAvatar from "@material-ui/core/ListItemAvatar";
import ListItemText from "@material-ui/core/ListItemText";

import { User } from "../../types/user";
import EditCommentIconButton from "./EditCommentIconButton";
import DeleteCommentIconButton from "./DeleteCommentIconButton";
import InputWrapper from "@component/inputWrapper";

interface VariablesUpdate {
  variables: {
    updateCommentInput: {
      commentId: string;
      description: string;
    };
  };
}

interface VariablesDelete {
  variables: {
    deleteCommentInput: {
      commentId: string;
      postId: string;
    };
  };
}

interface CommetProps {
  name: string;
  description: string;
  creatorId: string;
  userCommentedId: string;
  userConnected: User;
  postId: string;
  id: string;
  updateComment: (variables: VariablesUpdate) => any;
  deleteComment: (variables: VariablesDelete) => any;
  isClickEdit: boolean;
  isShown: boolean;
  setIsShown: (boolean) => void;
  setIsClickEdit: (boolean) => void;
}

// interface CommentType {
//   description: string;
//   userId: User;
//   postId: string;
//   id: string;
// }

// interface UpdatedCommetResponse {
//   updateComment: CommentType;
// }

// interface DeleteComment {
//   commentId: string;
//   postId: string;
// }

// interface DeletedCommetResponse {
//   deleteComment: DeleteComment;
// }

// interface CommentEditFormInput {
//   description: string;
//   commentId: string;
// }

// interface FormUpdate {
//   description: string;
//   commentId: string;
// }

// interface UpdateCommentInput {
//   updateCommentInput: FormUpdate;
// }

// interface DeleteCommentInput {
//   deleteCommentInput: DeleteComment;
// }

const Comment: React.FC<CommetProps> = (props: CommetProps) => {
  const {
    updateComment,
    deleteComment,
    name,
    description,
    userConnected,
    userCommentedId,
    creatorId,
    postId,
    id,
    isClickEdit,
    isShown,
    setIsShown,
    setIsClickEdit,
  } = props;
  const variables = { commentId: id, description };

  return (
    <>
      {isClickEdit ? (
        <InputWrapper
          label="Modifier ton commentaire"
          variables={variables}
          updateComment={updateComment}
          setIsClickEdit={setIsClickEdit}
        />
      ) : (
        <div
          id="comment-actions"
          style={{ display: "contents" }}
          onMouseEnter={(): void => setIsShown(true)}
          onMouseLeave={(): void => setIsShown(false)}
        >
          <ListItemAvatar>
            <Avatar alt="Profile Picture" src="" />
          </ListItemAvatar>
          <ListItemText primary={name} secondary={description} />
          {isShown && (
            <>
              {(userConnected.id === userCommentedId ||
                userConnected.id === creatorId) && (
                <DeleteCommentIconButton
                  deleteComment={deleteComment}
                  postId={postId}
                  commentId={id}
                />
              )}
              {userConnected.id === userCommentedId && (
                <EditCommentIconButton setIsClickEdit={setIsClickEdit} />
              )}
            </>
          )}
        </div>
      )}
    </>
  );
};

export default Comment;
