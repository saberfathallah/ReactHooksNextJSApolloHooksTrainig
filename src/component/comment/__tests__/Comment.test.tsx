import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import Comment from "../Comment";

describe("Comment", () => {
  const deleteComment = jest.fn();
  const updateComment = jest.fn();
  const setIsShown = jest.fn();
  const setIsClickEdit = jest.fn();

  let props = {
    deleteComment,
    setIsShown,
    updateComment,
    setIsClickEdit,
    isShown: true,
    isClickEdit: true,
    categoryId: "5e89c6458245cd0d75dec50f",
    name: "commment",
    description: "description",
    creatorId: "5e89c2fd8245cd0d75dec501",
    postId: "5e89c6878245cd0d75dec510",
    userCommentedId: "5e89c6878245cd0d75dec510",
    id: "5e89c6878245cd0d75dec510",
    userConnected: {
      id: "5e89c6878245cd0d75dec510",
      email: "sberrr@gmail.com",
      name: "sberrr",
    },
  };

  it("should match snapshot isShown===true && isClickEdit===true", () => {
    const tree = renderer.create(<Comment {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it("should match snapshot isShown===true && isClickEdit===false display edit and delete icons", () => {
    props.isClickEdit = false;

    const tree = renderer.create(<Comment {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it("should match snapshot isShown===true && isClickEdit===false display delete icon", () => {
    props.isClickEdit = false;
    props.userConnected.id = "5e89c2fd8245cd0d75dec501";

    const tree = renderer.create(<Comment {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it("should match snapshot isShown===false && isClickEdit===true", () => {
    props.isClickEdit = true;
    props.isShown = false;

    const tree = renderer.create(<Comment {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it("should match snapshot isShown===false && isClickEdit===false", () => {
    props.isClickEdit = false;
    props.isShown = false;

    const tree = renderer.create(<Comment {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it("should call setIsShown", () => {
    props.isClickEdit = false;
    props.isShown = false;

    const wrapper = mount(<Comment {...props} />);
    wrapper.find("#comment-actions").simulate("mouseenter");
    expect(setIsShown).toHaveBeenCalledWith(true);
    wrapper.find("#comment-actions").simulate("mouseleave");
    expect(setIsShown).toHaveBeenCalledWith(false);
  });

  it("should call deleteComment", () => {
    props.isShown = true;

    const wrapper = mount(<Comment {...props} />);
    wrapper.find("button").simulate("click");
    expect(deleteComment).toHaveBeenCalledTimes(1);
  });
});
