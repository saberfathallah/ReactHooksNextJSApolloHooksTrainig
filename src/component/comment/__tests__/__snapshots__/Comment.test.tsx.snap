// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment should match snapshot isShown===false && isClickEdit===false 1`] = `
<div
  id="comment-actions"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "display": "contents",
    }
  }
>
  <div
    className="MuiListItemAvatar-root"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiAvatar-fallback"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
        />
      </svg>
    </div>
  </div>
  <div
    className="MuiListItemText-root MuiListItemText-multiline"
  >
    <span
      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
    >
      commment
    </span>
    <p
      className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
    >
      description
    </p>
  </div>
</div>
`;

exports[`Comment should match snapshot isShown===false && isClickEdit===true 1`] = `
<form
  onSubmit={[Function]}
  style={
    Object {
      "display": "contents",
    }
  }
>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    style={
      Object {
        "margin": 8,
      }
    }
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
      data-shrink={true}
      htmlFor="descriptionstandard-full-width"
      id="descriptionstandard-full-width-label"
    >
      Modifier ton commentaire
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        disabled={false}
        id="descriptionstandard-full-width"
        name="description"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder=""
        required={false}
        type="description"
        value="description"
      />
    </div>
  </div>
  <button
    aria-label="modify"
    className="MuiButtonBase-root MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
</form>
`;

exports[`Comment should match snapshot isShown===true && isClickEdit===false display delete icon 1`] = `
<div
  id="comment-actions"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "display": "contents",
    }
  }
>
  <div
    className="MuiListItemAvatar-root"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiAvatar-fallback"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
        />
      </svg>
    </div>
  </div>
  <div
    className="MuiListItemText-root MuiListItemText-multiline"
  >
    <span
      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
    >
      commment
    </span>
    <p
      className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
    >
      description
    </p>
  </div>
  <button
    aria-label="delete"
    className="MuiButtonBase-root MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`Comment should match snapshot isShown===true && isClickEdit===false display edit and delete icons 1`] = `
<div
  id="comment-actions"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "display": "contents",
    }
  }
>
  <div
    className="MuiListItemAvatar-root"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiAvatar-fallback"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
        />
      </svg>
    </div>
  </div>
  <div
    className="MuiListItemText-root MuiListItemText-multiline"
  >
    <span
      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
    >
      commment
    </span>
    <p
      className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
    >
      description
    </p>
  </div>
  <button
    aria-label="delete"
    className="MuiButtonBase-root MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
  <button
    aria-label="modify"
    className="MuiButtonBase-root MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`Comment should match snapshot isShown===true && isClickEdit===true 1`] = `
<form
  onSubmit={[Function]}
  style={
    Object {
      "display": "contents",
    }
  }
>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    style={
      Object {
        "margin": 8,
      }
    }
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink={true}
      htmlFor="descriptionstandard-full-width"
      id="descriptionstandard-full-width-label"
    >
      Modifier ton commentaire
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        disabled={false}
        id="descriptionstandard-full-width"
        name="description"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder=""
        required={false}
        type="description"
        value="description"
      />
    </div>
  </div>
  <button
    aria-label="modify"
    className="MuiButtonBase-root MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
    </span>
  </button>
</form>
`;
